src/meadowflow/time_event_publisher.py

```diff
@@ -684,82 +716,45 @@ def call_at(self, dt: datetime.datetime, callback: Callable[[], None]) -> None:
-            else:
-                prev_next_callback = self._callback_queue[0]
-
-            # add the callback
-            heapq.heappush(self._callback_queue, time_and_callback)
-
-            # If we've changed when the next callback is, we need to notify the main
-            # loop. new_next_callback == prev_next_callback doesn't require a
-            # notification--the run function will just pick up whichever one happens to
-            # be in front and execute that one
-            if prev_next_callback is None or time_and_callback < prev_next_callback:
-                # we have to call Event.set on its event loop for it to have the desired
-                # effect. We currently don't assume that we are on self._event_loop in
-                # this function so we use call_soon_threadsafe
-                self._event_loop.call_soon_threadsafe(
-                    lambda: self._queue_modified.set()
-                )
-
-    async def main_loop(self) -> None:
-        """
-        Runs forever.
+        if len(self._callback_queue) == 0:
+            prev_next_callback = None
+        else:
+            prev_next_callback = self._callback_queue[0]
 
-        If main_loop is called more than once, subsequent calls will have no effect.
+        # add the callback
+        heapq.heappush(self._callback_queue, time_and_callback)
 
-        TODO the concurrency model needs to be clarified for this class as well as
-         Scheduler and EventLog. Currently, the model is roughly that there are two
-         threads, the main thread where a user "pokes at the system" (e.g. adds
-         callbacks), and a background thread running "the" event loop which updates the
-         system in the background. This is not at all rigorous though and probably
-         should be changed
+        # If we've changed when the next callback is, we need to notify the main
+        # loop. new_next_callback == prev_next_callback doesn't require a
+        # notification--the run function will just pick up whichever one happens to
+        # be in front and execute that one
+        if prev_next_callback is None or time_and_callback < prev_next_callback:
+            asyncio.get_running_loop().call_soon(self._queue_modified.set)
 
-        Callbacks run on "the" event loop, so they cannot be blocking.
+    async def _call_main_loop(self) -> None:
         """
+        Runs forever.
 
-        if asyncio.get_running_loop() != self._event_loop:
-            raise ValueError(
-                "_CallAt.main_loop was called from a different _event_loop than "
-                "expected"
-            )
+        Callbacks run on "the" event loop, so they cannot be blocking.
 
-        # Avoid running more than once. No lock is necessary as long as this always gets
-        # run on self._event_loop...
-        if self._is_running:
-            return
-        self._is_running = True
+        This coroutine is started as part of initialization when this class is
```

Comment:
Comment formatting--"awaited" can fit on the previous line :) I don't know why I notice these...